pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,844 @@
+
+				// Create or find function
+				funcId := s.findOrAddFunction(profile, nameIdx, filenameIdx, line.LineNumber)
+
+				profile.Location[locIdx].Line[j] = &googlev1.Line{
+					FunctionId: funcId,
+					Line:       int64(line.LineNumber),
+				}
+			}
+		}
+	}
+
+	mapping.HasFunctions = true
+	mapping.HasFilenames = true
+	mapping.HasLineNumbers = true
+}
+
+// findOrAddString finds a string in the string table or adds it if not found
+func (s *ProfileSymbolizer) findOrAddString(profile *googlev1.Profile, str string) int64 {
+	for i, s := range profile.StringTable {
+		if s == str {
+			return int64(i)
+		}
+	}
+
+	idx := int64(len(profile.StringTable))
+	profile.StringTable = append(profile.StringTable, str)
+	return idx
+}
+
+// findOrAddFunction finds a function in the function table or adds it if not found
+func (s *ProfileSymbolizer) findOrAddFunction(profile *googlev1.Profile, nameIdx, filenameIdx int64, lineNumber uint64) uint64 {
+	maxFuncId := uint64(0)
+	for _, f := range profile.Function {
+		maxFuncId = max(maxFuncId, f.Id)
+		if f.Name == nameIdx && f.Filename == filenameIdx {
+			return f.Id
+		}
+	}
+
+	funcId := maxFuncId + 1
+	profile.Function = append(profile.Function, &googlev1.Function{
+		Id:        funcId,
+		Name:      nameIdx,
+		Filename:  filenameIdx,
+		StartLine: int64(lineNumber),
+	})
+
+	return funcId
+}
+
+func (s *ProfileSymbolizer) Symbolize(ctx context.Context, req *Request) error {
+	start := time.Now()
+	status := StatusSuccess
+	defer func() {
+		s.metrics.profileSymbolization.WithLabelValues(status).Observe(time.Since(start).Seconds())
+	}()
+
+	if s.checkSymbolCache(req) {
+		return nil
+	}
+
+	if s.checkLidiaTableCache(ctx, req) {
+		return nil
+	}
+
+	if s.checkObjectStoreCache(ctx, req) {
+		return nil
+	}
+
+	return s.fetchAndCacheFromDebuginfod(ctx, req, &status)
+}
+
+// symbolizeWithTable processes all locations in a request using a lidia table
+func (s *ProfileSymbolizer) symbolizeWithTable(_ context.Context, table *lidia.Table, ei *BinaryLayout, req *Request) error {
+	// Buffer for reusing in symbol lookups
+	var framesBuf []lidia.SourceInfoFrame
+
+	for _, loc := range req.Locations {
+		resolveStart := time.Now()
+		addr, err := MapRuntimeAddress(loc.Address, ei, Mapping{
+			Start:  loc.Mapping.Start,
+			Limit:  loc.Mapping.Limit,
+			Offset: loc.Mapping.Offset,
+		})
+		if err != nil {
+			return fmt.Errorf("normalize address: %w", err)
+		}
+
+		// Look up the address directly in the lidia table
+		frames, err := table.Lookup(framesBuf, addr)
+		if err != nil {
+			level.Error(s.logger).Log(
+				"msg", "failed to resolve address on Lidia table lookup",
+				"addr", fmt.Sprintf("0x%x", addr),
+				"binary", req.BinaryName,
+				"build_id", req.BuildID,
+				"error", err,
+			)
+			loc.Lines = s.createNotFoundSymbols(req.BinaryName, loc, addr)
+			s.metrics.debugSymbolResolution.WithLabelValues(StatusErrorServerError).Observe(time.Since(resolveStart).Seconds())
+			continue
+		}
```

Comment:
* We don't want update to metrics on each resolved location: it's expensive and not useful: me are measuring memory latency here.
* We don't want to write a log message on each failed `Lookup`. it will kill the system.